flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/expressions/converter/ExpressionConverter.java

```diff
@@ -0,0 +1,322 @@
+				return relBuilder.getRexBuilder().makeBigintLiteral(bigint);
+			case FLOAT:
+				//Float/Double type should be liked as java type here.
+				return relBuilder.getRexBuilder().makeApproxLiteral(
+						extractValue(valueLiteral, BigDecimal.class),
+						relBuilder.getTypeFactory().createSqlType(SqlTypeName.FLOAT));
+			case DOUBLE:
+				//Float/Double type should be liked as java type here.
+				return rexBuilder.makeApproxLiteral(
+						extractValue(valueLiteral, BigDecimal.class),
+						relBuilder.getTypeFactory().createSqlType(SqlTypeName.DOUBLE));
+			case DATE:
+				return relBuilder.getRexBuilder().makeDateLiteral(DateString.fromCalendarFields(
+						valueAsCalendar(extractValue(valueLiteral, java.sql.Date.class))));
+			case TIME_WITHOUT_TIME_ZONE:
+				return relBuilder.getRexBuilder().makeTimeLiteral(TimeString.fromCalendarFields(
+						valueAsCalendar(extractValue(valueLiteral, java.sql.Time.class))), 0);
+			case TIMESTAMP_WITHOUT_TIME_ZONE:
+				return relBuilder.getRexBuilder().makeTimestampLiteral(TimestampString.fromCalendarFields(
+						valueAsCalendar(extractValue(valueLiteral, java.sql.Timestamp.class))), 3);
+			case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+				TimeZone timeZone = TimeZone.getTimeZone(((FlinkContext) ((FlinkRelBuilder) this.relBuilder)
+						.getCluster().getPlanner().getContext()).getTableConfig().getLocalTimeZone());
+				return this.relBuilder.getRexBuilder().makeTimestampWithLocalTimeZoneLiteral(
+						new TimestampWithTimeZoneString(
+								TimestampString.fromMillisSinceEpoch(
+										extractValue(valueLiteral, java.time.Instant.class).toEpochMilli()),
+								timeZone)
+								.withTimeZone(DateTimeUtils.UTC_ZONE)
+								.getLocalTimestampString(), 3);
+			case INTERVAL_YEAR_MONTH:
+				return this.relBuilder.getRexBuilder().makeIntervalLiteral(
+						BigDecimal.valueOf(extractValue(valueLiteral, Integer.class)),
+						new SqlIntervalQualifier(TimeUnit.YEAR, TimeUnit.MONTH, SqlParserPos.ZERO));
+			case INTERVAL_DAY_TIME:
+				return this.relBuilder.getRexBuilder().makeIntervalLiteral(
+						BigDecimal.valueOf(extractValue(valueLiteral, Long.class)),
+						new SqlIntervalQualifier(TimeUnit.DAY, TimeUnit.SECOND, SqlParserPos.ZERO));
+			default:
+				break;
+		}
+		Object object = extractValue(valueLiteral, Object.class);
+		if (object instanceof TimePointUnit) {
+			TimeUnit value = timePointUnitToTimeUnit((TimePointUnit) object);
+			return relBuilder.getRexBuilder().makeFlag(value);
+		} else if (object instanceof TimeIntervalUnit) {
+			TimeUnitRange value = intervalUnitToUnitRange((TimeIntervalUnit) object);
+			return relBuilder.getRexBuilder().makeFlag(value);
+		} else {
+			return relBuilder.literal(extractValue(valueLiteral, Object.class));
+		}
+	}
+
+	@Override
+	public RexNode visit(FieldReferenceExpression fieldReference) {
+		// We can not use inputCount+inputIndex+FieldIndex to construct field of calcite.
```

Comment:
@dawidwys @twalthr What do you think about this question? Should it be a bug of QueryOperationConverter? But I don't find any feasible way in RelBuilder. Maybe we should keep same order in AggregateQueryOperation or adjust the order of existed OperationTree? Both ways look hard...

Reply:
Yes, I think I also stumbled upon this problem myself. I did not figure out a good solution for it back then. But I was thinking about reordering the fields when converting the QueryOperation to RelNode. I am not sure how feasible that is.

Reply:
I think it makes sense to open a JIRA issue for this issue.

Reply:
I create https://issues.apache.org/jira/browse/FLINK-13902